mc/Mongo-SDAM.package/MongoServerSelection.class/instance/execute.st

```diff
@@ -0,0 +1,16 @@
+public
+execute
+	"Detect a monitor that satisfy the readPreference and localThreshold, signalling MongoSDAMServerSelectionTimeout if none is found."
+	
+	| startMS |
+	startMS := Time millisecondClockValue.
+	[
+		"TODO: don't use nil as answer..."
+		selectedMonitor := self readPreference accept: self.
+		selectedMonitor ifNotNil: [ :monitor | ^ monitor ].
+
+		client minHeartbeatFrequency wait.
```

Comment:
What if minHeartbeatFrequency > socket timeout?

Let's take a look at two examples. In both cases we have readPreference = 'primary' and the primary is black-holing (100% packet loss).

1. Synchronous processing (e.g. in a worker reading from a queue)

Eventually all workers will get stuck waiting for one heartbeat and then maybe process something that is old/retried. Meanwhile maybe more work gets queued than the system can handle (and a LIFO is infeasible).


2. Zinc REST interface

Imagine a client has a 2s time out before work is retried. Imagine we are using the multi-threaded ZnServer and Voyage. Now every request creates a thread and then is blocked on the MongoServerSelection. This will create many requests and once the primary is reachable again a lot of work is done for things that are not needed.



A language like go is solving it by passing a "context" from the application level to lower levels. A context can have a deadline (abort when not done by time X). I wonder how we can solve this here? Why does each look-up need to use an instance of MongoServerSelection?


I am especially concerned with the second case. During a network blip an image might get overloaded and stays in an overloaded state (as there is no way to shed work).

Reply:
For minHeartbeatFrequency, it's a constant: they define it's 500ms and can't change... https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#id55

Reply:
I read a bit on go contexts and I think I get your idea. But I wonder if you see something wrong in this (simpler?) solution that just starts a watchdog that stops execution on timeout (comes in valueWithin:onTimeout:):
~~~Smalltalk
| response | 
[	1000 timesRepeat: [ 'processing...' traceCr. 0.5 seconds wait ].
	response := 'HTTP 200: OK' ]
		valueWithin: 3 seconds
		onTimeout: [ response := 'HTTP 500: timeout error' ].
^ response.
~~~